### ISSUE TITLE
Cache Binaries

### ISSUE THREAD

**Issue by titusfortner** [titusfortner/webdrivers#29]
Ideally each session shouldn't need to make a network call if the driver has been recently updated. Alternately if there's a way to programmatically establish the correct driver for the browser version and not unnecessarily check for updates.

**Comment by seanlinsley** [comment]
It looks like this gem hooks into Selenium here:

https://github.com/SeleniumHQ/selenium/blob/cb7dd78b0a24214450dd70212168f81c0ad50f48/rb/lib/selenium/webdriver/chrome/driver.rb#L37-L41

With these patches:

https://github.com/titusfortner/webdrivers/blob/645308c1fd986919e03e413a8cbe78df6218cd23/lib/webdrivers/selenium.rb#L3-L7

So, in order to prevent unnecessary network requests, `update` needs to be restructured. https://github.com/titusfortner/webdrivers/blob/645308c1fd986919e03e413a8cbe78df6218cd23/lib/webdrivers/common.rb#L10-L31

It seems like this might cover all use cases without requiring network requests after the correct version has been installed:
```ruby
if correct_binary?
  Webdrivers.logger.debug "Expected webdriver version found"
  binary
elsif desired_version.nil?
  raise "Unable to find the latest version of #{file_name}; try downloading manually from #{base_url} and place in #{install_dir}"
else
  remove
  download
end
```

**Comment by titusfortner** [comment]
`#correct_binary` is still going to make the remote call though.

It's slightly trickier with the new chromedriver version matching, but I'm thinking a way to globally turn off auto updates probably needs to be supported, as well as checking the timestamp on the file and not making another network call if it is greater than a day (or user configurable period) old.

Haven't had a chance to look at the current open PR yet which will influence what this needs to look like.

**Comment by seanlinsley** [comment]
>#correct_binary is still going to make the remote call though.

Ah, yeah because `desired_version` falls back to `latest_version`.

Checking a timestamp to prevent the network requests could end up introducing bugs if the wait period coincides with a browser release. To work around that you could check the age of both the browser, and the driver.

**Comment by titusfortner** [comment]
There's two steps now for chrome. Matching versions and ensuring latest point release. Probably not a big deal to be a day behind on a point release by default, so long as the versions are properly matched

**Comment by jrochkind** [comment]
See also #4, another gotcha the network access leads to, involving Webmock. 

I understand the motivation for live checking for versions on every run in webdrivers. But for those who had been using chromedriver-helper without this feature without problems... I do think it would be nice if webdrivers should as an option provide a feature to work similar to chromedriver-helper, where it can use 'last known at time of release' versions without the network being involved. 

I'm also not sure what the implications of the network fetch are on CI runtimes on things like travis or circle-ci, perhaps I want to try to get travis to cache the location that webdrivers is going to be downloading to, hmm.

**Comment by kapoorlakshya** [comment]
@jrochkind I just submitted a PR that'll get you what you want - no network being involved. It'll allow you to pin down the desired version and limit the gem from updating the binary or performing any network checks until you:

1. Change the desired version
2. Or unset it

Please see this spec for an example: https://github.com/titusfortner/webdrivers/pull/49/files#diff-dbad66b82ddd9a9fdd3e819d2421f77fR46

**Comment by titusfortner** [comment]
implemented with #104
